bin/rpc/src/types.ts

```diff
@@ -63,19 +63,59 @@ export class RpcError extends Error {
   constructor(
     public code: number,
     message: string,
+    public data?: unknown,
   ) {
     super(message);
   }
 }
 
+export enum RpcErrorCode {
+  BlockUnavailable = 1,
+  WorkReportUnavailable = 2,
+  DASegmentUnavailable = 3,
+  Other = 0,
+}
+
 export interface DatabaseContext {
   blocks: BlocksDb;
   states: StatesDb<State & EnumerableState>;
 }
 
-export type RpcMethod<T, R> = (params: T, db: DatabaseContext, chainSpec: ChainSpec) => Promise<R>;
+export const withValidation = <P extends z.ZodType, R extends z.ZodType>(
+  method: (params: z.infer<P>, db: DatabaseContext, chainSpec: ChainSpec) => Promise<z.infer<R>>,
+  paramsSchema: P,
+  resultSchema: R,
+) => ({
```

Comment:
Just a matter of personal preference, but I'd consider moving the schema before the function, so that it's defined next to the actual parameters, like so:
```ts
withValidation(z.array(z.string()), z.null, async ([stringParamsArray]) => {
  // ...
  return null;
});
```

Reply:
done